### PR TITLE
Improve the error message when calling `Image.file` on web

### LINKED ISSUE #88772: Bad error message when calling`Image.file` on web
```console
======== Exception caught by image resource service ================================================
The following UnsupportedError was thrown resolving an image codec:
Unsupported operation: _Namespace

When the exception was thrown, this was the stack: 
dart:sdk_internal 5041:11                                             throw_
dart:sdk_internal 52632:17                                            get _namespacePointer
dart:sdk_internal 50535:28                                            _namespacePointer
dart:sdk_internal 50538:31                                            _dispatchWithNamespace
dart:sdk_internal 50653:23                                            open
```

### PR DESCRIPTION
**Pull request by tugorez**
# Improve the error message when calling `Image.file` on web

Fixes https://github.com/flutter/flutter/issues/88772

### Before:
image

### After: 
image

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by tugorez** [comment]
@yjbanov PTAL

File: packages/flutter/lib/src/widgets/image.dart [hunk 1]
```diff
@@ -464,7 +464,10 @@ class Image extends StatefulWidget {
     this.filterQuality = FilterQuality.low,
     int? cacheWidth,
     int? cacheHeight,
-  }) : image = ResizeImage.resizeIfNeeded(cacheWidth, cacheHeight, FileImage(file, scale: scale)),
+  }) :
+       // FileImage is not supported on Flutter Web therefore neither this method.
+       assert(!kIsWeb, 'Image.file is not supported on Flutter Web.'),
```
Comment (yjbanov) [inline packages/flutter/lib/src/widgets/image.dart:469]: The error message could offer `Image.asset` and `Image.network` as alternatives.

**Review by yjbanov — COMMENTED** [review]
Thanks for the contribution! As a test you could simply add one to `test/widgets/image_test.dart` but make it `skip: !kIsWeb`.

File: packages/flutter/lib/src/widgets/image.dart (hunk 1 above)
Comment (tugorez) [reply packages/flutter/lib/src/widgets/image.dart]: Great catch. Done. PTAL

**Comment by tugorez** [comment]
@yjbanov I think this PR is ready to be reviewed again. Thanks :)!

File: packages/flutter/test/widgets/image_test.dart [hunk 2]
```diff
@@ -1956,6 +1957,18 @@ void main() {
...
+    const String expectedError =
+      'Image.file is not supported on Flutter Web. '
+      'Consider using either Image.asset or Image.network instead.';
+    final Uri uri = Uri.parse('/home/flutter/dash.png');
+    final File file = File.fromUri(uri);
+    expect(
+      () => Image.file(file),
+      throwsA(predicate((AssertionError e) => e.message == expectedError)),
+    );
+  }, skip: !kIsWeb); // https://github.com/flutter/flutter/issues/88772 (Improve the error message when calling Image.file on web)
```
Comment (yjbanov) [inline packages/flutter/test/widgets/image_test.dart:1971]: Actually, I think I was wrong. Using `skip` this way goes against the spirit of the skip policy. The linked issue is meant to remind us to remove the `skip` in the future, but in this case the skip is intentional, we want it to stay forever. So instead, I propose a different approach: remove `skip` and instead test for different behaviors on different platforms as follows:

```dart
expect(
  () => Image.File(file),
  kIsWeb
    // Web does not support file access, expect AssertionError
    ? throwsA(predicate((AssertionError e) => e.message == expectedError))
    // AOT supports file access, expect constructor to succeed
    : isNot(throws),
),
```

Comment (tugorez) [reply packages/flutter/test/widgets/image_test.dart]: Done. PTAL

File: packages/flutter/lib/src/widgets/image.dart [hunk 3]
```diff
@@ -464,7 +464,13 @@ class Image extends StatefulWidget {
     this.filterQuality = FilterQuality.low,
     int? cacheWidth,
     int? cacheHeight,
-  }) : image = ResizeImage.resizeIfNeeded(cacheWidth, cacheHeight, FileImage(file, scale: scale)),
+  }) :
+       // FileImage is not supported on Flutter Web therefore neither this method.
+       assert(
+          !kIsWeb,
+          'Image.file is not supported on Flutter Web. '
+          'Consider using either Image.asset or Image.network instead.'),
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/image.dart:472]: Style guide nits:
```suggestion
       assert(
         !kIsWeb,
         'Image.file is not supported on Flutter Web. '
         'Consider using either Image.asset or Image.network instead.',
        ),
```

For more guidance you can read about the Flutter style guide here: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

File: packages/flutter/test/widgets/image_test.dart [hunk 4]
```diff
@@ -1956,6 +1957,22 @@ void main() {
       matchesGoldenFile('image_test.missing.2.png'),
     );
   }, skip: kIsWeb); // https://github.com/flutter/flutter/issues/74935 (broken assets not being reported on web)
+
+   testWidgets('Image.file throws a non-implemented error on web', (WidgetTester tester) async {
```
Comment (Piinks) [inline packages/flutter/test/widgets/image_test.dart:1961]: Can you also update this test to reflect the style guide? Thanks!

**Review by Piinks — COMMENTED** [review]
Hi @tugorez welcome! And thanks for the contribution!

File: packages/flutter/test/widgets/image_test.dart (hunk 4 above)
Comment (tugorez) [reply packages/flutter/test/widgets/image_test.dart]: Done. PTAL

**Review by Piinks — APPROVED** [review]
LGTM! Thank you! @yjbanov can you give a secondary review?
I think the enforcement for the new policy just went into effect, or is about to any moment now. 😋

**Review by werainkhatri — APPROVED** [review]
LGTM.

**Pull request merged**